plugins/builds/update.js

```diff
@@ -245,16 +270,99 @@ module.exports = () => ({
 
-            // Guard against triggering non-successful or unstable builds
-            // Don't further trigger pipeline if intented to skip further jobs
+            // Update stageBuild status if it has changed;
+            // if stageBuild status is currently terminal, do not update
+            const stage = await getStage({
+                stageFactory,
+                workflowGraph: newEvent.workflowGraph,
+                jobName: job.name,
+                pipelineId: pipeline.id
+            });
+            const isStageTeardown = STAGE_TEARDOWN_PATTERN.test(job.name);
+            let stageBuildHasFailure = false;
+
+            if (stage) {
+                const stageBuild = await stageBuildFactory.get({
+                    stageId: stage.id,
+                    eventId: newEvent.id
+                });
 
+                if (stageBuild.status !== newBuild.status) {
+                    if (!TERMINAL_STATUSES.includes(stageBuild.status)) {
+                        stageBuild.status = newBuild.status;
+                        await stageBuild.update();
+                    }
+                }
+
+                stageBuildHasFailure = TERMINAL_STATUSES.includes(stageBuild.status);
+            }
+
+            // Guard against triggering non-successful or unstable builds
+            // Don't further trigger pipeline if intend to skip further jobs
             if (newBuild.status !== 'SUCCESS' || skipFurther) {
                 // Check for failed jobs and remove any child jobs in created state
                 if (newBuild.status === 'FAILURE') {
-                    await removeJoinBuilds({ pipeline, job, build: newBuild }, request.server.app);
+                    await removeJoinBuilds(
+                        { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                        request.server.app
+                    );
                 }
+                // Do not continue downstream is current job is stage teardown and statusBuild has failure
+            } else if (newBuild.status === 'SUCCESS' && isStageTeardown && stageBuildHasFailure) {
+                await removeJoinBuilds(
+                    { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                    request.server.app
+                );
             } else {
-                await triggerNextJobs({ pipeline, job, build: newBuild, username, scmContext }, request.server.app);
+                await triggerNextJobs(
+                    { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                    request.server.app
+                );
+            }
+
+            // Determine if stage teardown build should start
+            // (if stage teardown build exists, and stageBuild.status is negative,
+            // and there are no active stage builds, and teardown build is not started)
+            if (stage) {
```

Comment:
We can skip this block, if the `newBuild` status is not a terminal status
```suggestion
            if (stage && FINISHED_STATUSES.includes(newBuild.status)) {
```